Add `MessageDispatcher` to standardize background task event loop logic

criccomini opened the pull request:
We have many separate background tasks in SlateDB:

- `mem_table_flush.rs` flushes memtables, handles checkpoint creation (sometimes), and refreshes `Db`'s in-memory manifest.
- `db_reader.rs` refreshes the in-memory `DbReader`s manifest.
- `compactor.rs` schedules compactions and checks the executor channel for responses.
- `garbage_collector.rs` handles garbage collection.
- `batch_write.rs` receives write requests and sends them to the WAL buffer and active (in-memory) mem table.
- `wal_buffer.rs` receives write requests and writes to the WAL.

Each one of these files has a bespoke (and often _very_ complex) `spawn_*` method. Yet each of these methods are largely doing the same set of things:

1. Handling messages from a channel
2. Handling clock ticker ticks
3. Handling failures (updating `DbState::error`, draining remaining messages, notifying remaining listeners, and so on)

I find these `spawn_` methods very difficult to understand and debug. In 0.7.0, we had multiple issues that stemmed from mismanagement of these event loops and their channels.

This PR aims to remove some of this complexity by providing a standard event loop implementation called `MessageDispatcher`. The dispatcher is responsible for managing the event loop lifecycle. Dispatchers are _not_ responsible for processing logic; that responsibility is delegated to `MessageHandler`s, which implement callbacks to handle event processing, tickers, and resource cleanup.

agavra commented on slatedb/src/dispatcher.rs:
```diff
@@ -0,0 +1,813 @@
+
+/// A dispatcher that invokes [MessageHandler] callbacks when events occur.
+///
+/// [MessageDispatcher] receives a [MessageHandler] and optional
+/// [mpsc::UnboundedReceiver<T>]. Messages sent to the receiver are passed to the
```
is the `rx` optional? I remember seeing that in the original PR but I don't see that one here (and also the docs for `new` need to be updated if that's the case)

criccomini replied:
It was. I got rid of it. All loops except the GC require it, so I went with simplifying the constructor (now that we have fp_registry). I suppose the other option is a builder pattern, which seemed rather overkill. WDYT? I can add an Option<rx>, builder pattern, more constructors, or... :)

agavra replied:
🫡 I prefer it this way, let's just have the GC loop pass in a no-op one.

criccomini replied:
Fixed docs: 453a639895ff1fa659c2c9ca54034a96cd793796

agavra commented on slatedb/src/dispatcher.rs:
```diff
@@ -0,0 +1,813 @@
+        )
+    }
+
+    /// Creates a new [MessageDispatcher] with a channel. This is the primary way to
+    /// create a dispatcher. Messages sent to the channel are passed to the
+    /// [MessageHandler] for processing.
```
similarly it looks like this needs updating

agavra commented on slatedb/src/dispatcher.rs:
```diff
@@ -0,0 +1,813 @@
+            .collect::<Vec<_>>();
+        // stop if we're in an error state or cancelled
+        let mut check_shutdown = async move || {
+            tokio::select! {
+                _ = cancellation_token.cancelled() => {},
+                _ = error_watcher.await_value() => {},
+            }
```
should we log on clean cancellation vs. error?

agavra commented on slatedb/src/dispatcher.rs:
```diff
@@ -0,0 +1,813 @@
+    ///
+    /// ## Returns
+    ///
+    /// A [Result] containing [SlateDBError::BackgroundTaskShutdown] on clean shutdown,
```
(can be in a future PR if we agree to change it) this has always confused me a bit - any reason not to make it `Ok(BackgroundTaskShutdown)` or something like that? Why is this an "error"?

criccomini replied:
Opened: https://github.com/slatedb/slatedb/issues/817

agavra commented on slatedb/src/dispatcher.rs:
```diff
@@ -0,0 +1,813 @@
+        let join = tokio::spawn(async move { dispatcher.run().await });
+
+        assert_eq!(log.lock().unwrap().clone(), vec![]);
+        clock.advance(Duration::from_millis(5)).await;
```
should we test with two tickers that overlaps (two go off on the same tick?) e.g. [3,5] would overlap at ts=15

criccomini replied:
Will do.

criccomini replied:
Done! And I added one for identical ticks as well.

agavra reviewed (approved):
🥳 nice abstraction fun tests

rodesai reviewed (approved):
LGTM!

criccomini commented:
Ok! I've updated this PR with a bunch of test tweaks. The was a really nasty race condition in the tests because I was modifying the clock in the main test thread while the tickers were updating their clock in the event loop thread. This led to some cases where the tickers might see the clock either before or after the clock update. This resulted in random hanging tests.

I also changed the dispatcher keep unused ticker futures rather than recreating them in every loop. I was worried it might be losing ticks otherwise. This required changing `selsect_all` to `FuturesUnordered` (for lifetime reasons), which seems to have a nicer API anyway.

I've now run the tests in a tight loop 900+ times without failure (previously, failures always showed almost immediately).

I am going to merge this PR since I have a ✅ from @rodesai. LMK if anyone has any other issues.
